sklearn/linear_model/_sag_fast.pyx.tp

```diff
@@ -515,6 +518,22 @@ def sag{{name_suffix}}(SequentialDataset{{name_suffix}} dataset,
                                   fabs(weights[idx] -
                                        previous_weights[idx]))
                 previous_weights[idx] = weights[idx]
+
+            with gil:
+                if _eval_callbacks_on_fit_iter_end(
```

Comment:
How does the overhead of taking the GIL compare to early stopping directly using the `stopping_criteron`?

Reply:
It has an impact on performance for sure. But If we want to enable callbacks at this step of the fit there's no way around.
What we can do however is to check before entering the nogil part if the estimator has callbacks and execute this part only if it's the case. Let me try something like that. We might encounter the same issue as in https://github.com/scikit-learn/scikit-learn/pull/13389